onnxruntime/test/shared_lib/test_inference.cc

```diff
@@ -391,6 +391,112 @@ TEST(CApiTest, create_session_without_session_option) {
 }
 #endif
 
+TEST(CApiTest, get_allocator_cpu) {
+  Ort::SessionOptions session_options;
+  Ort::ThrowOnError(OrtSessionOptionsAppendExecutionProvider_CPU(session_options, 1));
+  Ort::Session session(*ort_env, NAMED_AND_ANON_DIM_PARAM_URI, session_options);
+  Ort::MemoryInfo info_cpu = Ort::MemoryInfo::CreateCpu(OrtAllocatorType::OrtArenaAllocator, OrtMemTypeDefault);
+  Ort::Allocator cpu_allocator(session, info_cpu);
+
+  int com_result = 0;
+  Ort::ThrowOnError(Ort::GetApi().CompareMemoryInfo(info_cpu, cpu_allocator.GetInfo(), &com_result));
+  ASSERT_EQ(com_result, 0);
+  void* p = cpu_allocator.Alloc(1024);
+  ASSERT_NE(p, nullptr);
+  cpu_allocator.Free(p);
+}
+
+#ifdef USE_CUDA
+TEST(CApiTest, get_allocator_cuda) {
+  Ort::SessionOptions session_options;
+  Ort::ThrowOnError(OrtSessionOptionsAppendExecutionProvider_CUDA(session_options, 0));
+  Ort::Session session(*ort_env, NAMED_AND_ANON_DIM_PARAM_URI, session_options);
+
+  Ort::MemoryInfo info_cuda("Cuda", OrtAllocatorType::OrtArenaAllocator, 0, OrtMemTypeDefault);
+  Ort::Allocator cuda_allocator(session, info_cuda);
+
+  int com_result = 0;
+  Ort::ThrowOnError(Ort::GetApi().CompareMemoryInfo(info_cuda, cuda_allocator.GetInfo(), &com_result));
+  ASSERT_EQ(com_result, 0);
+  void* p = cuda_allocator.Alloc(1024);
+  ASSERT_NE(p, nullptr);
+  cuda_allocator.Free(p);
+}
+#endif
+
+TEST(CApiTest, io_binding) {
+  Ort::SessionOptions session_options;
+  Ort::ThrowOnError(OrtSessionOptionsAppendExecutionProvider_CPU(session_options, 1));
+  Ort::Session session(*ort_env, MODEL_URI, session_options);
+
+  Ort::MemoryInfo info_cpu = Ort::MemoryInfo::CreateCpu(OrtAllocatorType::OrtArenaAllocator, OrtMemTypeDefault);
+
+  const std::array<int64_t, 2> x_shape = {3, 2};
+  std::array<float, 3 * 2> x_values  = {1.0f, 2.0f, 3.0f, 4.0f, 5.0f, 6.0f};
+  Ort::Value bound_x = Ort::Value::CreateTensor(info_cpu, x_values.data(), x_values.size(),
+    x_shape.data(), x_shape.size());
+
+  const std::array<float, 3 * 2> expected_y = {1.0f, 4.0f, 9.0f, 16.0f, 25.0f, 36.0f};
+  const std::array<int64_t, 2> y_shape = {3, 2};
+  std::array<float, 3 * 2> y_values;
+  Ort::Value bound_y = Ort::Value::CreateTensor(info_cpu, y_values.data(), y_values.size(),
+                                                y_shape.data(), y_shape.size());
+
+
+  Ort::IoBinding binding(session);
+  binding.BindInput("X", bound_x);
+  binding.BindOutput("Y", bound_y);
```

Comment:
If we only call `BindInput`, then we ort will generate the output internally and we can use GetOutputValues to access it, right?

Reply:
You need to bind all outputs that you are interested in.

---
In reply to: [463803413](https://github.com/microsoft/onnxruntime/pull/4646#discussion_r463803413) [](ancestors = 463803413)

Reply:
But if bind output, then I need to know the shape of it. As we discussed, for some model, the output shape is related to the inference result, it can not be known before inference. I thought your PR will also support bind input only?

Reply:
For the cases of dynamic shapes you use a different API, it is called BindOutputToDevice. Please, read the documentation and see if this is adequate.

---
In reply to: [464509336](https://github.com/microsoft/onnxruntime/pull/4646#discussion_r464509336) [](ancestors = 464509336)

Reply:
Thanks, it looks good to me. It will be great if you can also add some test code here :).

Reply:
The test code is there. Per our conversation offline.

---
In reply to: [465246211](https://github.com/microsoft/onnxruntime/pull/4646#discussion_r465246211) [](ancestors = 465246211)